### PR TITLE
[Feature] Support inet_aton function

### LINKED ISSUE #49664: [Function] inet_acton.
Like inet_aton function in MySQL, return the numeric value of an IPv4 address.
For example :
select inet_aton('192.168.1.168');
3232235944

Firstly, we can access converting ipv4 to numeric value. Then ipv6, and conversion between IP and numerical value.

### PR DESCRIPTION
**Pull request by lufeizxy**
## What I'm doing:

Support inet_aton function
For #49664

```Plain Text
mysql> select inet_aton('192.168.1.1'); 
+--------------------------------------+ 
| inet_aton('192.168.1.1')             | 
+--------------------------------------+ 
| 3232235777                           | 
+--------------------------------------+ 
```

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
    - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.4
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by lufeizxy** [comment]
@Seaven  @wangsimo0  Can you help review this PR? Thank you!

**Comment by stdpain** [comment]
CI: Run echo "::error::Please check the format of your pr body about the behavior change checkbox!"

File: be/src/exprs/inet_aton.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+#include "common/format_ip.h"
+
+namespace starrocks {
+
+    static inline bool try_parse_ipv4(const char* pos, int64& result_value) {
+        return parse_ipv4_whole(pos, reinterpret_cast<unsigned char*>(&result_value));
+    }
+
+    template <IPConvertExceptionMode exception_mode, typename ToColumn>
+    static ColumnPtr convert_to_ipv4(const ColumnPtr haystack_ptr, const UInt8Column::Container* null_map = nullptr) {
```
Comment (stdpain) [inline be/src/exprs/inet_aton.cpp:31]: we could using unary template implements this function

example in lengh:
```
DEFINE_UNARY_FN_WITH_IMPL(lengthImpl, str) {
    return str.size;
}
```

File: be/src/exprs/inet_aton.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+#include "column/array_column.h"
+#include "column/binary_column.h"
+#include "column/column_viewer.h"
+#include "common/format_ip.h"
+
+namespace starrocks {
+
+    static inline bool try_parse_ipv4(const char* pos, int64& result_value) {
+        return parse_ipv4_whole(pos, reinterpret_cast<unsigned char*>(&result_value));
```
Comment (stdpain) [inline be/src/exprs/inet_aton.cpp:27]: prefer a better implement.

File: be/src/common/format_ip.h [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+inline bool is_numeric_ascii(char c) {
+    /// This is faster than
+    /// return UInt8(UInt8(c) - UInt8('0')) < UInt8(10);
+    /// on Intel CPUs when compiled by gcc 8.
+    return (c>='0' && c<='9');
+}
+
+template<typename T, typename EOFfunction>
+requires(std::is_same<typename std::remove_cv<T>::type, char>::value)
+inline bool parse_ipv4(T * &src, EOFfunction eof, unsigned char* dst, int64_t first_octet = -1) {
```
Comment (stdpain) [inline be/src/common/format_ip.h:41]: T*

**Comment by lufeizxy** [comment]
@stdpain Thank you for your feedback! I will address these suggestions as soon as possible.

**Comment by lufeizxy** [comment]
namespace starrocks {

static inline bool try_parse_ipv4(const char* pos, int64& result_value) {
    return parse_ipv4(pos, result_value);
}

DEFINE_UNARY_FN_WITH_IMPL(convert_to_ipv4_impl, str) {
    int64_t result;
    if (!try_parse_ipv4(str.data, result)) {
        return -1;
    }
    return result;
}

StatusOr<ColumnPtr> StringFunctions::inet_aton(FunctionContext* context, const Columns& columns) {
    
    return VectorizedStrictUnaryFunction<convert_to_ipv4_impl>::evaluate<TYPE_VARCHAR, TYPE_BIGINT>(column);
}

} // namespace starrocks
This implementation requires replacing -1 in the final result with null. Aside from modifying the evaluate method and validating input parameters on the FE side, I am not aware of any other efficient approach, so I have changed the implementation method. The other points mentioned in the PR have also been addressed.
@stdpain Could you please review the PR again? Thank you!

File: be/src/common/format_ip.h [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+    int64_t result = 0;
+    int offset = (IPV4_BINARY_LENGTH - 1) * IPV4_OCTET_BITS;
+
+    if (first_octet >= 0) {
+        result |= first_octet << offset;
+        offset -= IPV4_OCTET_BITS;
+    }
+
+    for (; true; offset -= IPV4_OCTET_BITS, ++pos) {
```
Comment (stdpain) [inline be/src/common/format_ip.h:47]: We need to check pos length.

Comment (lufeizxy) [reply be/src/common/format_ip.h:47]: ...
size_t length = std::strlen(pos);
if (length < 7 || length > 15) { 
     return false;
}

for (; true; offset -= IPV4_OCTET_BITS, ++pos) {
...
@stdpain Do you think this approach effectively addresses the concern you mentioned?I'm open to any feedback or additional insights you might have.Thank you!

Comment (stdpain) [reply be/src/common/format_ip.h:47]: the length should be str_value.get_size()
```
        auto str_value = str_viewer.value(row);
        int64_t parsed_result;
        if (try_parse_ipv4(str_value.get_data(), str_value.get_size(), parsed_result)) {
```

Comment (lufeizxy) [reply be/src/common/format_ip.h:51]: @stdpain I have modified the code according to your suggestions and tested it locally, confirming that everything runs smoothly. Could you please review the PR again? Thank you!

**Review by stdpain — DISMISSED** [review]

**Comment by lufeizxy** [comment]
@stdpain I will resolve it as soon as possible.

**Comment by lufeizxy** [comment]
@stdpain I have resolved the CI failures in my cases. Could you please review the PR again? Thank you!
image

**Review by stdpain — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by kevincai — APPROVED** [review]

**Pull request merged**